### PR TITLE
Add Support for Triggering Events via `AnimationEvent`s

### LINKED ISSUE #15494: animation triggers/events
## What problem does this solve or what need does it fill?

Being able to run arbitrary code at a defined point in time on an animation is sometimes really useful.

### Example use case 

- spawn particles/sound effects every time a characters feet hits the ground in a walk animation. 
- a character is attacking, deal damage to nearby enemies after a delay.

## What solution would you like?

I'd like a way to define "triggers" in an animations that executes a command or similar (event?), rather than animating some property.

something like:

```rust
#[derive(Event, Clone)]
enum Say {
    Hello,
    Bye,
}

fn say(trigger: Trigger<Say>) {
    match trigger.event() {
        Say::Hello => println!("Helloo!"),
        Say::Bye => println!("Bye!"),
    }
}

animation.trigger(1.0, Say::Hello); // can be any event type
animation.trigger(2.0, Say::Bye);
```

or maybe:

```rust
animation.add_triggers([1.0, 2.0], ["hello", "bye"]);
```

## What alternative(s) have you considered?

You can get pretty far using `Timer`s but it would take some effort to have it synced with an animation.

## Additional context

Also seen in other engines such as [unreal engine](https://dev.epicgames.com/documentation/en-us/unreal-engine/animation-notifies-in-unreal-engine), [unity](https://docs.unity3d.com/Manual/script-AnimationWindowEvent.html) and [godot](https://docs.godotengine.org/en/latest/tutorials/animation/animation_track_types.html#call-method-track).

### PR DESCRIPTION
**Pull request by atornity**
# Objective

Add support for events that can be triggered from animation clips. This is useful when you need something to happen at a specific time in an animation. For example, playing a sound every time a characters feet hits the ground when walking.

Closes #15494 

## Solution

Added a new field to `AnimationClip`: `events`, which contains a list of `AnimationEvent`s. These are automatically triggered in `animate_targets` and `trigger_untargeted_animation_events`. 

## Testing

Added a couple of tests and example (`animation_events.rs`) to make sure events are triggered when expected.

---

## Showcase

`Events` need to also implement `AnimationEvent` and `Reflect` to be used with animations.

```rust
#[derive(Event, AnimationEvent, Reflect)]
struct SomeEvent;
```

Events can be added to an `AnimationClip` by specifying a time and event.

```rust
// trigger an event after 1.0 second
animation_clip.add_event(1.0, SomeEvent);
```

And optionally, providing a target id.

```rust
let id = AnimationTargetId::from_iter(["shoulder", "arm", "hand"]);
animation_clip.add_event_to_target(id, 1.0, HandEvent);
```

I modified the `animated_fox` example to show off the feature.

CleanShot 2024-10-05 at 02 41 57

### CODE REVIEW

**Comment by mweatherley** [comment]
This is cool! I'm pretty focused on trying to make sure that #15530 gets through expediently, but I'll take a close look at this soon.

File: crates/bevy_animation/src/lib.rs
Comment (mweatherley) [inline crates/bevy_animation/src/lib.rs:1329]: Nit: This could probably have a better name than `new` considering how it's being constructed.

File: crates/bevy_animation/src/lib.rs
Comment (mweatherley) [inline crates/bevy_animation/src/lib.rs:1338]:
```suggestion
        // Return early if the animation finished on a previous tick.
```

**Review by mweatherley — APPROVED** [review]
Some nitpicks and not much else. I'd prefer if someone else who's more well-versed in events/observers gave this a look as well before merging it.

File: crates/bevy_animation/src/animation_event.rs
Comment (mweatherley) [reply crates/bevy_animation/src/animation_event.rs:35]: Weird. Looks like this got screwed up somehow. It was a grammar fix `does -> do` on the following line.

File: crates/bevy_animation/src/animation_event.rs
Comment (atornity) [reply crates/bevy_animation/src/animation_event.rs:45]: yeah me to, lol

**Comment by mweatherley** [comment]
Note to self:  Never pull in the middle of reviewing, since it apparently screws up all the comment line counts.

File: crates/bevy_animation/src/lib.rs
Comment (atornity) [reply crates/bevy_animation/src/lib.rs:290]: good idea

File: crates/bevy_animation/src/animation_event.rs
Comment (mweatherley) [inline crates/bevy_animation/src/animation_event.rs:36]:
```suggestion
/// do not have to implement `clone_value`.
```

File: crates/bevy_animation/src/lib.rs
Comment (mweatherley) [inline crates/bevy_animation/src/lib.rs:1335]: Nit: This could probably have a better name than `new` considering how it's being constructed.

**Review by mweatherley — APPROVED** [review]

File: examples/animation/animation_events.rs
Comment (atornity) [reply examples/animation/animation_events.rs:32]: That's a good idea, I'll leave that for a follow up tough

**Review by ItsDoot — APPROVED** [review]
For the ECS side of things (event triggering specifically), I like the design of this.

This isn't an approval for the animation side of things, however, as I don't have a lot of knowledge in that department.

File: crates/bevy_animation/src/lib.rs
Comment (atornity) [reply crates/bevy_animation/src/lib.rs:719]: added `set_seek_time`, which also set `last_seek_time` so events in between won't be triggered. Also linked to and explained the differences in the docs

File: crates/bevy_animation/src/lib.rs
Comment (atornity) [reply crates/bevy_animation/src/lib.rs]: renamed to `from_animation`

**Comment by mweatherley** [comment]
Another random nit: The `bunny_detective` example lets you spam spacebar to make it keep jumping, and I think it would feel more polished if it made you wait until after the bunny landed.

**Comment by atornity** [comment]
> Another random nit: The `bunny_detective` example lets you spam spacebar to make it keep jumping, and I think it would feel more polished if it made you wait until after the bunny landed.

hehe I though it was funny to spam it. I'll change it though

**Comment by ItsDoot** [comment]
Here's an alternative approach if interested.

There's a second part of observer targeting that this doesn't currently use: `ComponentId` targets. For example, these are used in the engine by `OnAdd`, `OnInsert`, and `OnRemove` to let you filter by what component you want to deal with:

```rust
fn on_add(trigger: Trigger<OnAdd, MyComponent>) {
    // I can handle the MyComponent getting added to the entity
}
```

If the current approach's `AnimationEvent`s don't carry additional information and are just used as a marker, we could instead use them as `ComponentId` targets for a single generic `AnimationEvent` struct:

```rust
// We can require a separate `AnimationTrigger` type to be implemented for them, but it isn't strictly necessary to make it work.
// I would recommend it, though.
#[derive(Component, Reflect)]
#[reflect(Component)] // We need to be able to grab the ComponentId when triggering events
struct CauseBubbles;
```

To register them, it remains the same:

```rust
// trigger an event after 1.0 second
animation_clip.add_event(1.0, CauseBubbles);
// or targeted
let id = AnimationTargetId::from_iter(["shoulder", "arm", "hand"]);
animation_clip.add_event_to_target(id, 1.0, CauseBubbles);
```

But the difference is when we listen for the triggers:

```rust
fn animate_bubbles(trigger: Trigger<AnimationEvent, CauseBubbles>) {
    // The animation event provides the target, time, etc
    // But we know this event was triggered for what we registered it for
}
```

To trigger the events from the world, you'll need to implement a `TriggerTargets` that allows you to target entities and components at the same time, and you'll trigger them by grabbing the `ComponentId`s registered to the animation, and passing that to `World::trigger_targets`.

Let me know what you think!

File: crates/bevy_animation/src/lib.rs
Comment (ChristopherBiscardi) [inline crates/bevy_animation/src/lib.rs:1374]:
```suggestion
        // Return early if the animation have finished on a previous tick.
```

**Comment by atornity** [comment]
> But yeah, I think I'll stick with the current implementation for now, thanks for the suggestion though!

I should probably update the examples to make it clear that adding data to events is possible.

**Comment by mockersf** [comment]
I would prefer to not add a new model, would it be possible to work with the animated fox in examples?

if not, the model should be added to https://github.com/bevyengine/bevy/blob/main/CREDITS.md with its licence and a link to the opengameart.org page

**Comment by atornity** [comment]
> I would prefer to not add a new model, would it be possible to work with the animated fox in examples?
> 
> if not, the model should be added to https://github.com/bevyengine/bevy/blob/main/CREDITS.md with its licence and a link to the opengameart.org page

That seems doable. would you prefer I added events to the existing `animated_fox` example in that case? It would be two mostly identical examples otherwise, although the `animated_fox` example would become quite large if I did that.

**Comment by atornity** [comment]
I removed the `bunny_detective` example and added animation events to the `animated_fox` example instead.

File: examples/animation/animated_fox.rs
Comment (ChristopherBiscardi) [inline examples/animation/animated_fox.rs:42]:
```suggestion
    // The observer is not required to be defined in an impl block
    fn observer(
```

Small suggestion since there is no strong connection between the observer function and the type (ie: it doesn't take self, etc so you can't call it on an OnStep value, its for code organization). Newer programmers may find this pattern confusing and think it *has* to be in an impl block because of a lack of understanding the derive/reflects in play.

Alternatively it could be defined as a standalone function, for example: `fn observe_on_step`.

**Review by ChristopherBiscardi — APPROVED** [review]
Overall love this feature, I've written implementations in my games for it in the past. I ran the examples (including the older ones that got removed) and feel like adding events is very usable.

I'm curious if the difference between manual impls and derived impls running at different times will be a pain point for learning the feature from examples.

One thing I wonder about is timeframe vs frame count for add_event. Will the time offset scale if time is sped up/slowed down (such as for a slow-motion effect)? and how does someone figure out what the time offset is supposed to be? If I was building an animation in another program wouldn't it be more likely that I was calculating it based on the frame data? Perhaps time and frame versions of specifying could be useful.

**Comment by atornity** [comment]
> I'm curious if the difference between manual impls and derived impls running at different times will be a pain point for learning the feature from examples.

Yep, it was a bad idea on my part to make an example that required a custom implementation since it modifies `Text`. Although it did _work_ with the derive, it just annoyed me that the color would change one frame before the text. On the other hand, it does demonstrate how to manually implement the trait, witch is more powerful anyways.

> One thing I wonder about is timeframe vs frame count for add_event. Will the time offset scale if time is sped up/slowed down (such as for a slow-motion effect)? and how does someone figure out what the time offset is supposed to be? If I was building an animation in another program wouldn't it be more likely that I was calculating it based on the frame data? Perhaps time and frame versions of specifying could be useful.

In the bunny example I had a comment that explained how to determine the correct time, it's essentially `frame index / fps of animation` (regardless of playback speed), I'll see if I can add that to the fox example as well. 
You can use up and down keys in the fox example btw, to see it's still working when playing in slow motion or sped up. 

As for an alternate `add_event` version where you specify the frame the event should occur. That sounds super useful to me. I don't think I'll be adding that in this PR though as I'm not sure how to retrieve (or if it's possible) the frame time of an animation.

File: examples/animation/animated_fox.rs
Comment (atornity) [reply examples/animation/animated_fox.rs]: good idea, removed in favor of `observe_on_step`.

File: examples/animation/animated_fox.rs
Comment (mockersf) [inline examples/animation/animated_fox.rs:280]:
```suggestion
```

**Review by mockersf — APPROVED** [review]

**Comment by alice-i-cecile** [comment]
Thank you to everyone involved with the authoring or reviewing of this PR! This work is relatively important and needs release notes! Head over to https://github.com/bevyengine/bevy-website/issues/1716 if you'd like to help out.

**Pull request merged**
